src/webgpu/api/operation/memory_sync/buffer/multiple_buffers.spec.ts

```diff
@@ -27,9 +42,57 @@ g.test('rw')
     Perform a 'read' operations on multiple buffers, followed by a 'write' operation.
     Operations are separated by a 'boundary' (pass, encoder, queue-op, etc.).
     Test that the results are synchronized.
-    The read should not see the contents written by the subsequent write.`
+    The read should not see the contents written by the subsequent write.
+  `
   )
-  .unimplemented();
+  .params(u =>
+    u //
+      .combine('boundary', kOperationBoundaries)
+      .expand('_context', p => kBoundaryInfo[p.boundary].contexts)
+      .expandWithParams(function* ({ _context }) {
+        for (const readOp of kAllReadOps) {
+          for (const writeOp of kAllWriteOps) {
+            if (checkOpsValidForContext([readOp, writeOp], _context)) {
+              yield {
+                readOp,
+                readContext: _context[0],
+                writeOp,
+                writeContext: _context[1],
+              };
+            }
+          }
+        }
+      })
+  )
+  .fn(async t => {
+    const { readContext, readOp, writeContext, writeOp, boundary } = t.params;
+    const helper = new OperationContextHelper(t);
+
+    const srcBuffers: GPUBuffer[] = [];
+    const dstBuffers: GPUBuffer[] = [];
+
+    const kSrcBufferCount = 4;
+    for (let i = 0; i < kSrcBufferCount; i++) {
+      const { srcBuffer, dstBuffer } = await t.createBuffersForReadOp(readOp, kSrcValue, kOpValue);
+
+      srcBuffers.push(srcBuffer);
+      dstBuffers.push(dstBuffer);
+
+      await t.createIntermediateBuffersAndTexturesForWriteOp(writeOp, 0, kOpValue);
+
+      // The read op will read from src buffer and write to dst buffer based on what it reads.
+      // The write op will write the given op value into src buffer as well.
+      // The write op happens after read op. So we are expecting the src value to be in the dst
+      // buffer.
+      t.encodeReadOp(helper, readOp, readContext, srcBuffers[i], dstBuffers[i]);
+      helper.ensureBoundary(boundary);
+    }
+
+    t.encodeWriteOp(helper, writeOp, writeContext, srcBuffers[0], 0, kOpValue);
```

Comment:
@kainino0x It seems to me that unimplemented tests in the multiple_buffers.spec.ts can be implemented by referring to single_buffer.spec.ts. So, I'm trying to implement `rw` test first.  But, I'm not sure if I understand the description of this test correctly. I understood that this test should perform a read operation on multiple buffers, then perform a write operation on each buffer, right? It's likely I don't find the right way to write this test. Any comment?

Reply:
re: https://github.com/gpuweb/cts/pull/1857

The `memory_sync` tests are very difficult because they attempt to find implementations in memory synchronization primitives, like barriers issued to Vulkan to ensure a write completes before a read starts. This is the reason they've sat unimplemented for so long. Fortunately, I we have already written most of the `memory_sync,buffer` tests, and *think* the idea here was pretty straightforward - just add some complexity relative to the `single_buffer` tests to make the usage a little more robust, for example, to make sure barriers are placed on the correct resources.

I'll leave a separate comment about the test content.

BTW, the issue for this is https://github.com/gpuweb/cts/issues/897 (found by searching memory-sync in the project tracker: https://github.com/orgs/gpuweb/projects/3/views/1?filterQuery=is%3Aissue+-status%3AComplete%2CBlocked%2CStarted%2CPost-V1+memory_sync), so if you'll be working on this please assign yourself and mark it as started. This would have also given me a chance to look at the test and make sure it wasn't going to be too gnarly. For example, the `memory_sync,texture` tests have several extra layers of complexity over the `memory_sync,buffer` tests, and I want someone with some experience with Vulkan synchronization to write them (which excludes me, because I don't have much experience with it either).

Reply:
Thanks for explaining the `memory_sync` buffer tests. If this PR is ready to get reviews, let me take the issue(#897) because I was not sure if I can make this PR ready to get reviewed.